File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        
        vector<int> temp;
        int len = 0;
        for (int i = 0; i < n; i++) {
            int idx = lower_bound(temp.begin(), temp.end(), nums[i]) - temp.begin();
            if (idx == temp.size()) {
                temp.push_back(nums[i]);
                len++;
            } else {
                temp[idx] = nums[i];
            }
        }
        
        return len;
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
		
        dp = [1]*n
        maxLen = 1
        for i in range(n):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)       
            maxLen = max(maxLen, dp[i])
		
        return maxLen